### ISSUE TITLE
calico-node CrashLoopBackoff with termination reason completed

### ISSUE THREAD

**Issue by thomaspurchas** [projectcalico/calico#5157]
I'm seeing pretty much identical behaviour to #2822, `calico-node` is exiting after 1 second with exit code 0, and pod being marked as complete. There's nothing obvious in the logs to explain this behaviour.

## Expected Behavior
`calico-node` shouldn't exit.

## Current Behavior
I have node who's `calico-node` always exits with status code 0, pod deletions and reconfigurations have done nothing to convince it to stay alive.

```
Containers:
  calico-node:
    Container ID:   containerd://7968fa86125411f39cfe2b882b66094cdbce39e6029f2d5ac7a0e8e0dcfa33df
    Image:          docker.io/calico/node:v3.21.1
    Image ID:       docker.io/calico/node@sha256:a56f71d9f877c4bb75e2823c30c9bec3b1e29eae0ba97bc0dd01b25c69099cca
    Port:           <none>
    Host Port:      <none>
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Tue, 07 Dec 2021 00:26:09 +0000
      Finished:     Tue, 07 Dec 2021 00:26:19 +0000
    Ready:          False
    Restart Count:  8
    Requests:
      cpu:      250m
    Liveness:   exec [/bin/calico-node -felix-live] delay=10s timeout=10s period=10s #success=1 #failure=6
    Readiness:  exec [/bin/calico-node -felix-ready] delay=0s timeout=10s period=10s #success=1 #failure=3
    Environment Variables from:
      kubernetes-services-endpoint  ConfigMap  Optional: true
    Environment:
      DATASTORE_TYPE:                     kubernetes
      WAIT_FOR_DATASTORE:                 true
      NODENAME:                            (v1:spec.nodeName)
      CALICO_NETWORKING_BACKEND:          <set to the key 'calico_backend' of config map 'calico-config'>  Optional: false
      CLUSTER_TYPE:                       k8s,bgp
      IP:                                 autodetect
      CALICO_IPV4POOL_IPIP:               Never
      CALICO_IPV4POOL_VXLAN:              Always
      FELIX_IPINIPMTU:                    <set to the key 'veth_mtu' of config map 'calico-config'>  Optional: false
      FELIX_VXLANMTU:                     <set to the key 'veth_mtu' of config map 'calico-config'>  Optional: false
      FELIX_WIREGUARDMTU:                 <set to the key 'veth_mtu' of config map 'calico-config'>  Optional: false
      CALICO_DISABLE_FILE_LOGGING:        true
      FELIX_DEFAULTENDPOINTTOHOSTACTION:  ACCEPT
      FELIX_IPV6SUPPORT:                  false
      FELIX_HEALTHENABLED:                true
      IP_AUTODETECTION_METHOD:            interface=tailscale.*
```

## Steps to Reproduce (for bugs)
Not really sure, my other node running what I think is the exact same config has no issues.

## Context
`calico-node` keeps crash looping causing all the pods on the node to be restarted. The VXLAN network overlay still works fine (I can ping pods on other nodes, from the unhealthy node), while `calico-node` is crash looping.

## Your Environment

* Calico version

```
node1:~# ./calicoctl version
Client Version:    v3.21.1
Git commit:        183af6a8
Cluster Version:   v3.21.1
Cluster Type:      k8s,bgp,kubeadm,kdd
```

* Orchestrator version (e.g. kubernetes, mesos, rkt):

`v1.22.4`

* Operating System and version:

Debian Bullseye (11.0)

## Debug logs from crash looping `calico-node`

https://gist.github.com/thomaspurchas/40361d289857b05acb3562959106aeac
```
...
2021-12-07 00:26:15.861 [DEBUG][63] felix/health.go 131: New health report lastReport=health.HealthReport{Live:true, Ready:true} name="int_dataplane" newReport=&health.HealthReport{Live:true, Ready:true}
2021-12-07 00:26:15.861 [DEBUG][63] felix/table.go 1234: In nftables mode, restarting transaction between updates and deletions. ipVersion=0x4 table="filter"
2021-12-07 00:26:15.861 [DEBUG][63] felix/table.go 1266: Update ended up being no-op, skipping call to ip(6)tables-restore. ipVersion=0x4 table="filter"
2021-12-07 00:26:15.861 [DEBUG][63] felix/health.go 131: New health report lastReport=health.HealthReport{Live:true, Ready:true} name="int_dataplane" newReport=&health.HealthReport{Live:true, Ready:true}
2021-12-07 00:26:15.861 [DEBUG][63] felix/health.go 131: New health report lastReport=health.HealthReport{Live:true, Ready:true} name="int_dataplane" newReport=&health.HealthReport{Live:true, Ready:true}
2021-12-07 00:26:15.861 [DEBUG][63] felix/int_dataplane.go 1884: Asked to reschedule. delay=3.942642161s
2021-12-07 00:26:15.861 [INFO][63] felix/summary.go 100: Summarising 1 dataplane reconciliation loops over 0s: avg=28ms longest=28ms (resync-filter-v4)
... dropped 1 logs ...
2021-12-07 00:26:19.029 [DEBUG][63] felix/health.go 148: GET /readiness
2021-12-07 00:26:19.029 [DEBUG][63] felix/health.go 182: Checking state of reporter reporter=&health.reporterState{name:"felix-startup", reports:health.HealthReport{Live:true, Ready:true}, timeout:0, latest:health.HealthReport{Live:true, Ready:true}, timestamp:time.Time{wall:0xc063c868d346a84c, ext:195571689, loc:(*time.Location)(0x4a164c0)}}
2021-12-07 00:26:19.029 [DEBUG][63] felix/health.go 182: Checking state of reporter reporter=&health.reporterState{name:"int_dataplane", reports:health.HealthReport{Live:true, Ready:true}, timeout:20000000000, latest:health.HealthReport{Live:true, Ready:true}, timestamp:time.Time{wall:0xc063c869f35a7996, ext:4733741363, loc:(*time.Location)(0x4a164c0)}}
2021-12-07 00:26:19.029 [DEBUG][63] felix/health.go 182: Checking state of reporter reporter=&health.reporterState{name:"async_calc_graph", reports:health.HealthReport{Live:true, Ready:true}, timeout:20000000000, latest:health.HealthReport{Live:true, Ready:true}, timestamp:time.Time{wall:0xc063c868e3602aab, ext:465678920, loc:(*time.Location)(0x4a164c0)}}
2021-12-07 00:26:19.029 [INFO][63] felix/health.go 196: Overall health status changed newStatus=&health.HealthReport{Live:true, Ready:true}
2021-12-07 00:26:19.029 [DEBUG][63] felix/health.go 199: Calculated health summary healthResult=&health.HealthReport{Live:true, Ready:true}
2021-12-07 00:26:19.029 [DEBUG][63] felix/health.go 151: Health: ready
```

**Comment by caseydavenport** [comment]
Have you tried passing the `-p` argument to `kubectl logs` to get the logs from the previous container run? That might have some more clues.

**Comment by thomaspurchas** [comment]
After digging around and `strace`ing the crap out of everything, I have figured out that the pod is being sent a `SIGKILL`, from somewhere. Still trying to figure out why it’s being sent a `SIGKILL`, the usual suspect of `OOM` is staying very quiet, nothing in system logs to suggest a an `OOM`.

The `SIGKILL` gives you an 0 exit code because `calico-node` uses `runsvdir` to start everything up, it unhelpful exits with a 0 exit code with it’s sent a `SIGKILL`. It might be worth considering a different service runner, or wrapping `runsvdir` with something that returns a 139 exit code on `SIGKILL` to avoid confusing exits like this. 

http://manpages.ubuntu.com/manpages/bionic/man8/runsvdir.8.html

**Comment by caseydavenport** [comment]
@thomaspurchas normally I'd suspect the kubelet sending that signal due to liveness probes failing, but the logs seems to suggest that Calico considers itself "alive" - perhaps the kubelet liveness checks are failing to talk to Calico and so assumes that Calico is not running properly?

**Comment by thomaspurchas** [comment]
I've tried setting the `failureThreshold` of the readiness and liveliness checks to `999`, but still seeing the same behaviour. However I have noticed a pod event that stands out `Pod sandbox changed, it will be killed and re-created.`. As far as I can tell, this happens after felix gets going and starts configuring network interfaces etc. I'm wondering if it's somehow reconfiguring the interfaces on it's own sandbox, and causing `kubelet` to then restart it. It comes back up, reconfigures again, gets killed again etc etc.

I'm kinda reaching here, my knowledge of both k8s and calico is far too limited to investigate this hypothesis properly. I don't suppose you have any hints @caseydavenport?

**Comment by ldumont** [comment]
I think I have the same issue, I see the calico-node pod in CrashLoopBackoff on Debian 11.

Actually I'm testing the upgrade of my cluster from Debian 10 to Debian 11 so I have one master and one node running on Debian 10 and I added a new node running Debian 11. And the CrashLoopBackoff is only on the new Debian 11 node

On Debian 10 I'm using iptables-legacy but on the new node I'm using iptables-nft. Using iptables-legacy on Debian 11 doesn't change anything.

There is maybe something related to the operating system ?

kubernete 1.22.5
calico 3.21.2
debian 11

**Comment by caseydavenport** [comment]
> However I have noticed a pod event that stands out Pod sandbox changed, it will be killed and re-created.

This is most likely unrelated, or at least unlikely to be the cause - calico/node runs in the host netns so doesn't receive a dedicated sandbox. That log is likely referring to another pod-networked pod on the machine. 

> 2021-12-07 00:26:11.753 [DEBUG][63] felix/health.go 131: New health report lastReport=health.HealthReport{Live:true, Ready:false} name="int_dataplane" newReport=&health.HealthReport{Live:true, Ready:false}

This log seems to indicate something is making calico unhappy and consider itself not ready. These logs look a bit suspicious to me:

```
2021-12-07 00:26:11.610 [WARNING][63] felix/endpoint_mgr.go 1175: Could not set accept_ra: <nil> ifaceName="cali3879dac72fe"
```

Unfortunately looks like the error there is masked, so I've raised a PR to fix that diag: https://github.com/projectcalico/calico/pull/5317/

Are you running with any special settings? e.g., in non-privileged mode?

**Comment by caseydavenport** [comment]
> Actually I'm testing the upgrade of my cluster from Debian 10 to Debian 11 so I have one master and one node running on Debian 10 and I added a new node running Debian 11. And the CrashLoopBackoff is only on the new Debian 11 node


@ldumont could be that calico/node isn't properly detecting the iptables version on the host. It normally auto-detects which mode to use (legacy or nft) based on the presence of rules, but it's not foolproof. Might be worth checking to see if its writing to the correct iptables module.

**Comment by thomaspurchas** [comment]
> Are you running with any special settings? e.g., in non-privileged mode?

Not using any special settings at the moment. I’ve now nuked the original cluster and rebuilt it. It’s a tad more stable, but still seeing the `calico-nodes` periodically get stuck in a `CrashLoopBackoff` for 2 or 3 restarts, then they’re stable again. Also noticed this is only happening on non-controller nodes. The Kube controller node has shown no issues so far.

> I'd expect this to be the case - try setting it explicitly to NFT and see if that fixes it. Debian 11 should be using NFT, IIRC.

I’ll give this a shot as well and see what happens.

**Comment by caseydavenport** [comment]
@thomaspurchas any luck on this? 

At a minimum, we'll be releasing v3.22 soon which will fix the bug masking the error in the log message you're seeing.

**Comment by ldumont** [comment]
I was testing a few components upgrades to see if it would change anything, when I setup the new debian11 node using Docker instead of Containerd and it worked ! I tried to go back to Containerd with the same exact setup and the problem is still there.

So it might be an issue with calico and Containerd ?

kubernetes: 1.23.3
Containerd: 1.4.12
Docker: 20.10.12
Iptables: 1.8.7 (nf_tables)

**Comment by caseydavenport** [comment]
Hm, possible that it's a runtime issue although not one that I've seen before and I know lots of folks use containerd with Calico.

**Comment by caseydavenport** [comment]
@thomaspurchas @ldumont did you guys ever figure out what was going on here?

**Comment by ldumont** [comment]
Hello, I think my issue was not related to Calico. 

I did more testing and I was strugguling to get a stable cluster even before deploying Calico. The issue seem to be that Debian 11 and containerd don't work well using systemd cgroup driver. 

As soon as I switched to cgroupfs in containerd and kubelet, the cluster was stable and I could depliy Calico without any issue. The cluster has been stable since.

Thanks for your help.
